File: packages/guards/src/chain-of-thought-leak.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ChainOfThoughtLeakOptions {
  action: 'block' | 'warn';
  /** Custom internal markers to detect */
  customMarkers?: string[];
}

const COT_PATTERNS = [
  /\b(?:let me think|let's think|step by step|chain of thought|reasoning:)\b/i,
  /\b(?:internal(?:ly)?|privately|behind the scenes)\b.*\b(?:think|reason|consider|analyze)\b/i,
  /<(?:thinking|scratchpad|internal|cot|reasoning)>/i,
  /\[(?:THINKING|SCRATCHPAD|INTERNAL|COT|REASONING)\]/i,
  /\b(?:my internal reasoning|my thought process|my analysis shows)\b/i,
  /^(?:Step \d+:|Phase \d+:|First,|Second,|Third,|Finally,)/im,
];

export function chainOfThoughtLeak(options: ChainOfThoughtLeakOptions): Guard {
  const patterns = [...COT_PATTERNS];
  if (options.customMarkers) {
    for (const m of options.customMarkers) {
      patterns.push(new RegExp(m.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'i'));
    }
  }

  return {
    name: 'chain-of-thought-leak',
    version: '0.1.0',
    description: 'Detects leaked chain-of-thought reasoning in output',
    category: 'security',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];
      for (const p of patterns) {
        const m = text.match(p);
        if (m) matched.push(m[0]);
      }
      const triggered = matched.length > 0;
      return {
        guardName: 'chain-of-thought-leak', passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matched: [...new Set(matched)] } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/chain_of_thought_leak.py
"""Detects chain-of-thought reasoning leaks in output."""
from __future__ import annotations

import re
import time
from typing import List

from open_guardrail.core import GuardResult

_PATTERNS = [
    re.compile(r"<thinking>", re.IGNORECASE),
    re.compile(r"\[SCRATCHPAD\]", re.IGNORECASE),
    re.compile(r"let\s+me\s+think\s+step\s+by\s+step", re.IGNORECASE),
    re.compile(r"my\s+internal\s+reasoning", re.IGNORECASE),
    re.compile(r"Step\s+1\s*:", re.IGNORECASE),
    re.compile(r"<internal>", re.IGNORECASE),
]


class _ChainOfThoughtLeak:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "chain-of-thought-leak"
        self.action = action

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matched: List[str] = []
        for p in _PATTERNS:
            m = p.search(text)
            if m:
                matched.append(m.group())
        triggered = len(matched) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="chain-of-thought-leak",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"CoT leak detected: {', '.join(matched)}" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"matched": matched} if triggered else None,
        )


def chain_of_thought_leak(*, action: str = "block") -> _ChainOfThoughtLeak:
    return _ChainOfThoughtLeak(action=action)
